src/main.rs

```diff
@@ -1,3 +1,72 @@
+extern crate pancurses;
+
+use pancurses::Window;
+use pancurses::{initscr, endwin, Input, noecho};
+
 fn main() {
-    println!("Hello, world!");
+    let window = initscr();
+    window.refresh();
+    window.keypad(true);
+    noecho();
+    render_line((0, 0), (40, 10), &window);
+    loop {
+        match window.getch() {
+            Some(Input::Character('q')) => break,
+            _ => ()
+        }
+    }
+    endwin();
+}
+
+trait RenderOutput {
+    fn set_pixel(&self, x: i32, y: i32);
+    fn get_size(&self) -> (i32, i32);
+}
+
+impl RenderOutput for Window {
+    fn set_pixel(&self, x: i32, y: i32) {
+        self.mvaddch(y, x, '#');
+    }
+
+    fn get_size(&self) -> (i32, i32) {
+        (self.get_max_x(), self.get_max_y())
+    }
+}
+
+fn render_line(p0: (i32, i32), p1: (i32, i32), out: &RenderOutput) {
```

Comment:
I'm getting a warning that `RenderOutput` should have a `dyn` on it.  Any idea what that's about?

Reply:
## Dynamic Dispatch and Trait Objects

Yes `dyn` is quite confusing! It's very similar to, in Java, having a function accept not a concrete type but "a thing that implements an interface". This is "dynamic dispatch" - you are saying that `render_line` will take a reference to _something_ that implements `RenderOutput`. That "something that implements `RenderOutput`" is called a "trait object". A trait object is a couple of things:

1. a pointer to the actual item in memory
2. a pointer to the item's vtable that points to the functions that implement the functions required by the specified trait

That way, at runtime, Rust can follow these pointers to get the correct implementations of the function. The reason why Rust requires the `dyn` (_I think_) is to make the dynamic dispatch element _explicit_ (since it's a performance hit most people don't want to take). That being said, what you have is fine until you see fit to change it :-)

## Alternatives

The only alternative I know would be to make `render_line` _generic_. That is, have it accept "some concrete type that implements `RenderOutput`" (which is _slightly_ different than "a reference to some unspecified type that implements `RenderOutput`"). The difference in syntax would be one of the following three options:

```rust
fn render_line<T: RenderOutput>(p0: ..., p1: ..., out: &T) { }
fn render_line(p0: ..., p1: ..., out: &impl RenderOutput) { }
fn render_line<T>(p0: ..., p1: ..., out: &T)
    where T: RenderOutput
{ }
```

These three all result in the same thing.

## What's the Difference?

The difference is incredibly non obvious at this point. The change really occurs doing compilation where Rust performs [monomorphization](https://doc.rust-lang.org/book/ch10-01-syntax.html?highlight=monomorph#performance-of-code-using-generics). Basically, when using generics, Rust will look at your code at compilation time, find _all the places_ where `render_line` is called, look at the _concrete type_ passed to `render_line`, and then (assuming that concrete type implements `RenderOutput`) generate a _specific implementation_ of `render_line` for that type. For example if you had:

```rust
trait Foo {}
impl Foo for i8 {}
impl Foo for i16 {}
fn generic_fn<T: Foo>(t: T) { }
fn main() {
    let x: i8 = 0;
    generic_fn(x);
    let y: i16 = 0;
    generic_fn(y);
}
```

The compiled code will look something along the lines of:

```rust
fn generic_fn_i8(t: i8) {}
fn generic_fn_i16(t: i16) {}
fn main() {
    let x: i8 = 0;
    generic_fn_i8(x);
    let y: i16 = 0;
    generic_fn_i16(y);
}
```

This allows for more performant implementations of the `generic_fn` function since Rust can perform optimizations specific to the type being called.

## If It's Slower, When Should I Use It?

Rust can't monomorphize implementations of generic code if it can't identify the concrete type being passed when the function is called. For instance, if you want to have an array of `i8`s and `i16`s and you want to call `generic_fn` on each item Rust won't know which item is an `i8` and which item is an `i16` at runtime so it _must_ dynamically dispatch.

## Moar Resources

A _great_ resource is [the Rust Book](https://doc.rust-lang.org/book/ch17-02-trait-objects.html) but a possibly more engaging and in-depth explanation comes from the wonderful [Jon Gjengset](https://www.youtube.com/watch?v=xcygqF5LVmM).

Reply:
This was a great explanation!

I ended up going with the generic version since I felt the monomorphization was just a bit more _rusty_ (albeit to my neophyte eyes).